### PR TITLE
[WIP][BEAM-9379] Update calcite to 1.26

### PR DESCRIPTION
**Pull request by nielsbasjes**
In these commits I have done the following:
1. Update the vendored calcite build to version 1.25
1. Use that new Calcite version in Beam.
1. Attempt (!!) to fix all the changes that have occurred in Calcite over the last year.

To test this I have run these to install the new calcite locally 
```
./gradlew -p vendor/calcite-1_25_0  check
./gradlew -p vendor/calcite-1_25_0  -PvendoredDependenciesOnly  publishToMavenLocal 
```

Current state of this initial pull request:
1. WORK IN PROGRESS. And I have currently reached the boundaries of my knowledge of this part of the system. Help is needed.
1. It probably won't build in CI because the vendored Calcite has not yet been released.
1. There are lots of changes that I did to make it compile that I'm not sure are correct. My current knowledge of this is too limited at this time.
1. If I run this locally (i.e. ./gradlew build -p sdks/java/extensions/sql/  ) it all builds and about 10% of the tests fail (i.e. about 90% pass)

R:@amaliujia

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [x] Update `CHANGES.md` with noteworthy changes.
 - [x] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Dataflow | Flink | Samza | Spark | Twister2
--- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/) | ---
Java | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/)
Python | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python38/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/) | ---
XLang | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/) | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website | Whitespace | Typescript
--- | --- | --- | --- | --- | --- | ---
Non-portable | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocker_Cron/lastCompletedBuild/) <br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocs_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Whitespace_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Typescript_Cron/lastCompletedBuild/)
Portable | --- | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | --- | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by nielsbasjes** [comment]
R: @amaliujia

**Comment by kennknowles** [comment]
Can you publish a build scan by running your second step with `--scan` ? Then we can see what is broken.

**Comment by kennknowles** [comment]
It might be a huge amount of work. We can just add the new vendored module and then release it, but we will want to know that we have enough people available to also port to it.

**Comment by vectorijk** [comment]
> It might be a huge amount of work. We can just add the new vendored module and then release it, but we will want to know that we have enough people available to also port to it.

agree on this. we could vendor module first and use released library for porting work.

last few weeks. I did a testing on my branch https://github.com/vectorijk/beam/tree/calcite-1-25 and run
`./gradlew -Ppublishing -PvendoredDependenciesOnly -PjavaLinkageArtifactIds=beam-vendor-calcite-1_25_0:0.1-SNAPSHOT :checkJavaLinkage`
it still has some issues unresolved.

File: vendor/calcite-1_25_0/build.gradle [hunk 1]
```diff
@@ -18,29 +18,33 @@
 
 plugins { id 'org.apache.beam.vendor-java' }
 
-description = "Apache Beam :: Vendored Dependencies :: Calcite 1.20.0"
+description = "Apache Beam :: Vendored Dependencies :: Calcite 1.25.0"
 
 group = "org.apache.beam"
 version = "0.2"
```
Comment (vectorijk) [inline vendor/calcite-1_25_0/build.gradle:24]: let's start from 0.1?

File: vendor/calcite-1_25_0/build.gradle [hunk 2]
```diff
@@ -18,29 +18,33 @@
...
         "org.apache.http",
+        "org.apiguardian.api",
         "org.codehaus",
+        "org.objectweb",
         "org.pentaho",
-        "org.yaml"
+        "org.yaml",
 ]
 
 vendorJava(
```
Comment (vectorijk) [inline vendor/calcite-1_25_0/build.gradle:50]: some logging libraries need to be put into runtimeDepencies. for you reference: https://github.com/vectorijk/beam/blob/calcite-1-25/vendor/calcite-1_25_0/build.gradle

File: vendor/calcite-1_25_0/build.gradle (hunk 1 above)
Comment (nielsbasjes) [reply vendor/calcite-1_25_0/build.gradle:24]: > Can you publish a build scan by running your second step with `--scan` ? Then we can see what is broken.

https://scans.gradle.com/s/ofcfnu774du4y

Comment (nielsbasjes) [reply vendor/calcite-1_25_0/build.gradle:24]: Yes

**Comment by nielsbasjes** [comment]
> Can you publish a build scan by running your second step with `--scan` ? Then we can see what is broken.

https://scans.gradle.com/s/ofcfnu774du4y

File: vendor/calcite-1_25_0/build.gradle (hunk 2 above)
Comment (nielsbasjes) [reply vendor/calcite-1_25_0/build.gradle:50]: I have copied (+ updated version) of the runtime dependencies part of this.
As this is uncharted territory for me I'm unsure if this is correct now.

**Comment by kennknowles** [comment]
https://scans.gradle.com/s/ofcfnu774du4y/tests/:sdks:java:extensions:sql:test/org.apache.beam.sdk.extensions.sql.BeamSqlDslSqlStdOperatorsTest/testArithmeticOperator#1

Looks like Calcite switched to `java.math.BigDecimal`?

**Comment by kennknowles** [comment]
https://scans.gradle.com/s/ofcfnu774du4y/tests/:sdks:java:extensions:sql:test/org.apache.beam.sdk.extensions.sql.BeamSqlDslUdfUdafTest/testUdfWithListInput#1

Seems parsing of arrays changed?

**Comment by kennknowles** [comment]
https://scans.gradle.com/s/ofcfnu774du4y/tests/:sdks:java:extensions:sql:zetasql:test/org.apache.beam.sdk.extensions.sql.zetasql.ZetaSqlTimeFunctionsTest/testFormatDate#1

this one looks easier - just a change of capital to lowercase CC @robinyqiu

**Comment by kennknowles** [comment]
https://scans.gradle.com/s/ofcfnu774du4y/tests/:sdks:java:extensions:sql:test/org.apache.beam.sdk.extensions.sql.impl.rel.BeamMatchRelTest/matchPrevFunctionTest#1

Seems there is something special and missing from `BeamMatchRel`

**Comment by kennknowles** [comment]
https://scans.gradle.com/s/ofcfnu774du4y/tests/:sdks:java:extensions:sql:test/org.apache.beam.sdk.extensions.sql.BeamSqlDslAggregationTest/testWindowOnNonTimestampField#1

This is just a bad test. It checks the exact string instead of the key elements of the string so it is sensitive to irrelevant changes.

**Comment by kennknowles** [comment]
https://scans.gradle.com/s/ofcfnu774du4y/tests/:sdks:java:extensions:sql:test/org.apache.beam.sdk.extensions.sql.impl.udf.BeamSqlUdfExpressionTest/testSHA1?anchor=e30&focused-exception-line=0-1-0#1

Switch from `byte[]` to avatica's `ByteString`

**Comment by iemejia** [comment]
Can we consider moving this one to 1.26 because of https://lists.apache.org/thread.html/r0b0fbe2038388175951ce1028182d980f9e9a7328be13d52dab70bb3%40%3Cannounce.apache.org%3E

Probably Beam's use case of Calcite is not impacted by this but I can envision the automatic vulnerability tools complaining on this soon.

**Comment by nielsbasjes** [comment]
I am in need of advice / assistance.

What I ran into is that some tests in the current state of this pull request fail over a change in Calcite.
I am specifically talking about https://github.com/apache/beam/blob/master/sdks/java/extensions/sql/jdbc/src/test/java/org/apache/beam/sdk/extensions/sql/jdbc/BeamSqlLineTest.java#L63 and many other tests in this file that do a `CREATE EXTERNAL TABLE` or a `DROP TABLE`.

I tracked the source of this exception back to Calcite where a few months ago this default method was added in Calcite by @julianhyde : https://github.com/apache/calcite/blob/master/core/src/main/java/org/apache/calcite/sql/parser/SqlParserImplFactory.java#L58

This default implementation simply returns a `DdlExecutor` (a new interface) that always fails with a `UnsupportedOperationException` if no valid implementation has been provided.
Looks good so far.

The problem I have is that my knowledge is lacking how to correctly override this method to return the implementation for Beam that makes it possible to do a `CREATE EXTERNAL TABLE`.

A pointer on how I should fix this correctly is highly appreciated.

**Comment by aaltay** [comment]
> I am in need of advice / assistance.

@tysonjh @amaliujia @kennknowles - Who could help on this PR?

**Comment by nielsbasjes** [comment]
Yes please. 
I have tried over the last week and the main problems that remain are beyond me.
I'm stuck and unable to proceed.
At this point I do not have the understanding on how Calcite and Beam (should) work together to fix this.
So I'm totally fine if someone takes over from here.

At this point I see these `big` categories of tests that fail:

1. Tests that try to create a table fail with `DDL not supported`.
2. Tests that have nested records fail with class cast exceptions in the `SchemaCoder::encode`. 
    To me it seems an error exists in  handling the nesting of schemas.
3. The `BeamCalcRel::castOutputTime` tries to call `java.time.LocalDate.ofEpochDay(long)` yet the underlying Calcite code goes searching for `java.time.LocalDate.ofEpochDay(java.lang.Integer)` which doesn't exist. 
    To me this seems an internal change of calcite how they choose data types.

**Comment by apilloud** [comment]
Thanks for all the work you've done on this so far! Hopefully you saw [my response to your email to dev@](https://lists.apache.org/thread.html/ra9bfbc1f6db11dbc3d033fdc0e1015110032b1a88ce20be0301cee3f%40%3Cdev.beam.apache.org%3E) about about the DDL issues. The other two things you are seeing are probably around changes that need to be made to the type translation code in [BeamCalcRel.java](https://github.com/apache/beam/blob/3bb232fb098700de408f574585dfe74bbaff7230/sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamCalcRel.java).

I don't have time to pick this up right now, but I do have time planned for this in Late January/Early February.

**Comment by nielsbasjes** [comment]
All of this was fixed in https://github.com/apache/beam/pull/14729

**Comment by apilloud** [comment]
Thanks for all the work you did to make this happen!

**Pull request closed without merging**